compiler/rustc_hir/src/def.rs

```diff
@@ -196,35 +209,126 @@ impl DefKind {
+/// - `s` will resolve to [`Res::Local`];
+/// - the call to `str_to_string` will resolve to [`Res::Def`], with the [`DefId`]
+///   pointing to the definition of `str_to_string` in the current crate.
+//
 #[derive(Clone, Copy, PartialEq, Eq, Encodable, Decodable, Hash, Debug)]
 #[derive(HashStable_Generic)]
 pub enum Res<Id = hir::HirId> {
+    /// Any ordinary definition not covered by the other `Res` variants.
+    ///
+    /// **Not bound to a specific namespace.**
     Def(DefKind, DefId),
 
     // Type namespace
+    /// A primitive type such as `i32` or `str`.
+    ///
+    /// **Belongs to the type namespace.**
     PrimTy(hir::PrimTy),
-    /// `Self`, with both an optional trait and impl `DefId`.
+    /// The `Self` type, optionally with the trait it is associated with
+    /// and optionally with the [`DefId`] of the impl it is associated with.
+    ///
+    /// **Belongs to the type namespace.**
+    ///
+    /// For example, the `Self` in
+    ///
+    /// ```
+    /// trait Foo {
+    ///     fn foo() -> Box<Self>;
+    /// }
+    /// ```
+    ///
+    /// would have the [`DefId`] of `Foo` associated with it. The `Self` in
+    ///
+    /// ```
+    /// struct Bar;
+    ///
+    /// impl Bar {
+    ///     fn new() -> Self { Bar }
+    /// }
+    /// ```
+    ///
+    /// would have the [`DefId`] of the impl associated with it. Finally, the `Self` in
+    ///
+    /// ```
+    /// impl Foo for Bar {
+    ///     fn foo() -> Box<Self> { Box::new(Bar) }
+    /// }
+    /// ```
+    ///
+    /// would have both the [`DefId`] of `Foo` and the [`DefId`] of the impl
+    /// associated with it.
+    ///
+    /// *See also [`Res::SelfCtor`].*
+    ///
+    /// -----
     ///
     /// HACK(min_const_generics): impl self types also have an optional requirement to not mention
     /// any generic parameters to allow the following with `min_const_generics`:
-    /// ```rust
+    /// ```
     /// impl Foo { fn test() -> [u8; std::mem::size_of::<Self>()] {} }
     /// ```
     /// We do however allow `Self` in repeat expression even if it is generic to not break code
     /// which already works on stable while causing the `const_evaluatable_unchecked` future compat lint.
     ///
     /// FIXME(lazy_normalization_consts): Remove this bodge once that feature is stable.
-    SelfTy(Option<DefId> /* trait */, Option<(DefId, bool)> /* impl */),
-    ToolMod, // e.g., `rustfmt` in `#[rustfmt::skip]`
+    SelfTy(
+        /// Optionally, the trait associated with this `Self` type.
+        Option<DefId>,
+        /// Optionally, the impl associated with this `Self` type.
+        Option<(DefId, bool)>,
+    ),
+    /// A tool attribute module; e.g., the `rustfmt` in `#[rustfmt::skip]`.
+    ///
+    /// **Belongs to the type namespace.**
+    ToolMod,
 
     // Value namespace
-    SelfCtor(DefId /* impl */), // `DefId` refers to the impl
+    /// The `Self` constructor, along with the [`DefId`]
+    /// of the impl it is associated with.
```

Comment:
Is this description accurate?

Reply:
No, constructors are not associated with an impl block. In the playground link I gave above, the `Y` that isn't re-exported to the crate root is the Ctor. I'm not sure if the DefId here is the id of the Ctor or of the type itself, petrochenkov may know.

Reply:
yeah, the constructor should contain the `DefId` of either the struct or the variant it is associated with i think.

They aren't really associated with impls

Reply:
Hmm, I grepped through the code and looked around a bit, and it seems that the `DefId` *is* from an `impl` block: https://github.com/rust-lang/rust/blob/8e6b478a5a146ad77ae4340ce379e1674c35ad69/compiler/rustc_resolve/src/late.rs#L1277

Reply:
That's not what that code is doing - it's adding `Self` to the value namespace for items within that impl block, which is what allows things like
```rust
struct S {}
impl S {
  fn f() -> S { ... }
}
```
But SelfCtor isn't limited to impl blocks, `struct S;` also generates a Ctor named `S`.

Reply:
`struct S;` generates a `Res::Def(DefKind::Ctor)` but IIUC `Res::SelfCtor` is **only** for `impl` blocks. The docs this thread is about are for `SelfCtor`.

Reply:
Ah ok, I didn't realize they were different. This seems fine then.